### PR TITLE
fix(encoding/base64url): throw TypeError in case of invalid base64url string

### LINKED ISSUE #1011: base64url decoder allows decoding base64
The encoding/base64url.ts decoder allows decoding base64 (not base64url) encoded strings.

image

The third case should error with the same error as the first case.

std@0.100.0 & deno 1.11.2 (release, x86_64-unknown-linux-gnu)

### PR DESCRIPTION
**Pull request by TotallyNotChase**
fixes #1011 

Validate input in base64url decoder.

Add tests for making sure invalid base64url strings are rejected.

### CODE REVIEW

File: encoding/base64url.ts [hunk 1]
```diff
@@ -17,6 +17,10 @@ export function addPaddingToBase64url(base64url: string): string {
 }
 
 function convertBase64urlToBase64(b64url: string): string {
+  if (!/^[-_A-Z0-9]*?={0,2}$/i.test(b64url)) {
+    // Contains characters not part of base64url spec.
+    throw new TypeError("Invalid base64url character in string!");
```
Comment (kt3k) [inline encoding/base64url.ts:22]: Maybe we should align the error message with `base64.decode` like:

```ts
throw new TypeError("Failed to decode base64url: invalid character);
```

---
BTW @lucacasonato do you think the error here should be DOMException?

**Comment by kt3k** [comment]
The test case looks good to me

File: encoding/base64url.ts (hunk 1 above)
Comment (TotallyNotChase) [reply encoding/base64url.ts]: Changed, for the time being. Awaiting input on whether or not this should be a `DOMException`.

**Review by bartlomieju — APPROVED** [review]
LGTM

**Pull request merged**